src/librustc_mir/diagnostics.rs

```diff
@@ -1991,6 +1991,25 @@ fn main() {
 ```
 "##,
 
+E0510: r##"
+Cannot mutate place in this match guard.
+
+When matching on a variable it cannot be mutated in the match guards, as this
+could cause the match to be non-exhaustive:
+
+```compile_fail
+let mut x = Some(0);
+match x {
+    None => (),
+    Some(v) if { x = None; false } => (),
+    Some(_) => (), // No longer matches
+}
+```
```

Comment:
Let's add a note below to make this clearer:

---

Here, executing the `x = None` statement in the match guard would modify the value being matched, requiring us to go "back in time" and match the `None` arm.